### PR TITLE
docs for using evented_dataclass

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
adds developer-focused documentation for using the `evented_dataclass` decorator.

@sofroniewn & @kevinyamauchi, particularly curious to hear if you feel like I've left any important details out based on your usage.  There's some subtleties to using `ClassVar` and `InitVar` that I haven't covered here (mostly not wanting to duplicate the existing python docs) ... but let me know if there are examples that you'd like me to add there.

### CODE REVIEW

File: docs/source/explanations/dataclass.md
Comment (sofroniewn) [inline docs/source/explanations/dataclass.md:206]: I havn't had to use this syntax yet (instead using `Property`) and I wonder if we expect it's usage to be so common that we need it in our docs. I don't mind, but I'm with you on trying to keep our docs as minimal as possible

Comment (tlambert03) [reply docs/source/explanations/dataclass.md:206]: which syntax? the default_factory?

Comment (sofroniewn) [reply docs/source/explanations/dataclass.md:206]: both `field` and `default_factory`, but maybe I've been using more tuples etc

Comment (tlambert03) [reply docs/source/explanations/dataclass.md:206]: I think it's important enough to mention.  If anyone wants to set an empty list as the default value for something, that's the way to do it... and _not_ `x: list = []`

Comment (sofroniewn) [reply docs/source/explanations/dataclass.md:206]: ok, good point

File: docs/source/explanations/dataclass.md
Comment (sofroniewn) [inline docs/source/explanations/dataclass.md:221]: `events=True` is default right? it might more confusing to see that called out if not necessary, but i don't mind it either

File: docs/source/explanations/dataclass.md
Comment (sofroniewn) [inline docs/source/explanations/dataclass.md:255]: similar about the default

File: docs/source/explanations/dataclass.md
Comment (sofroniewn) [inline docs/source/explanations/dataclass.md:257]: I find this such a useful part of the evented_dataclass that I might mention it in the intro about events too. While the original motivation wasn't to add this syntax I find it very clean and a big benefit

File: docs/source/explanations/dataclass.md
Comment (sofroniewn) [inline docs/source/explanations/dataclass.md:346]: These get called before the actual thing is set/ gotten right, might be nice to add that to make it extra clear.

File: docs/source/explanations/dataclass.md
Comment (sofroniewn) [inline docs/source/explanations/dataclass.md:391]: I might put this at the top of the special cases, or even in the main section as this is the syntax I used most commonly in our current evented dataclasses and I really like it

File: docs/source/explanations/dataclass.md
Comment (sofroniewn) [inline docs/source/explanations/dataclass.md:427]: Just a thought for the future, right now we don't do any actually type checking / enforcement with the type right? So in the example below event though we typed it as `int` if you passed a float you wouldn't get an error or it forced into an `int` would you? I wonder if that's something we could add as a bool to the `evented_dataclass`, like `strict=True` or something. I'm a newbie to typing in Python so this may be a little out of left field or just wrong.

**Review by sofroniewn — APPROVED** [review]
This is incredible @tlambert03 so helpful. I've gone through and left some comments, all are minor and optional. Nice work

File: docs/source/explanations/dataclass.md
Comment (tlambert03) [reply docs/source/explanations/dataclass.md:427]: correct, there's no runtime type enforcement.  As mentioned way back in https://github.com/napari/napari/issues/1436, there's a great library that does exactly this called [`pydantic`](https://github.com/samuelcolvin/pydantic/)... they even have their own [dataclass](https://pydantic-docs.helpmanual.io/usage/dataclasses/) that enforces types.  So we could either just use pydantic, or borrow/vendor the important bits.

Comment (sofroniewn) [reply docs/source/explanations/dataclass.md:427]: ah yes `pydantic` that was it! I remember at the time wanting to wait, but we might want to look into adding that now that we've started using the dataclass. Curious what @jni and others think. I also just realized I went to undergrad with the creator of pydantic, havn't seen him in > 10 years, not sure if he'd remember me!

Comment (tlambert03) [reply docs/source/explanations/dataclass.md:427]: > I also just realized I went to undergrad with the creator of pydantic

Ha! No way!
Yeah let me know if you want to try to incorporate type checking (or feel free to play with it yourself!)

File: docs/source/explanations/dataclass.md
Comment (kevinyamauchi) [reply docs/source/explanations/dataclass.md:206]: I agree that this is good to include! I've come across this myself.

File: docs/source/explanations/dataclass.md
Comment (kevinyamauchi) [inline docs/source/explanations/dataclass.md:338]: This section is super useful!

File: docs/source/explanations/dataclass.md
Comment (kevinyamauchi) [reply docs/source/explanations/dataclass.md:346]: Related question: are the `_on_{prop}_set()` methods called called to set the attribute values during the `__init__()`?

File: docs/source/explanations/dataclass.md
Comment (kevinyamauchi) [reply docs/source/explanations/dataclass.md:257]: I agree the `Property` type syntax is super nice.

I think it would be nice to add a sentence giving some guidance on using the `Property` vs. `_on_{prop}_set()` methods. As I understand things so far, it seems like the `Property` type is useful for simple type coercion and the `_on_{prop}_set()` methods are better when something more complex is required such as modifying other private attributes. However, I am curious as to what your thoughts are, @tlambert03 and @sofroniewn !

**Review by kevinyamauchi — APPROVED** [review]
I think this looks great! Thank you, @tlambert03 ! I left a couple of small comments below, but they are mostly clarifying questions for my own usage of the `evented_dataclass`

**Comment by sofroniewn** [comment]
@tlambert03 what are your thoughts on this? Do you want to make any more updates before merge? @kevinyamauchi and I left some comments, I don't think any are majorly major. This also doesn't include the `asdict` and `update` methods added in #1966, but they could be added later.

If we go with `pydantic` and the approach in #2042 then we'd no longer be using the `evented_dataclass` and so would need to rewrite this documentation, but there are still a variety of questions around that, and so it's probably worth getting this in now for `0.4.3`, but that might influence which if any changes you want to make before merge.

**Comment by tlambert03** [comment]
i don't know actually... I feel like it's more or less serving the purpose at the moment, which is just some thoughts that core developers might be interested in if they are converting things to `evented_dataclass`.  Given that it may not be what we ultimately use, I feel like it's fine to just leave it here and direct people to it as necessary, rather than merging just yet.

**Comment by sofroniewn** [comment]
> i don't know actually... I feel like it's more or less serving the purpose at the moment, which is just some thoughts that core developers might be interested in if they are converting things to evented_dataclass. Given that it may not be what we ultimately use, I feel like it's fine to just leave it here and direct people to it as necessary, rather than merging just yet.

Ok good points, agreed let's leave as is, and I will change the milestone from `0.4.3`. Thanks!

**Comment by sofroniewn** [comment]
Closing following #2042 and #2236 where we replaced the evented dataclass with the pydantic evented model

**Pull request closed without merging**
